PACTH-6: Add Shadow DOM test

yenleagest opened the pull request:
**Test case:**
1. Navigate to https://books-pwakit.appspot.com/
2. Search by "playwright"
3. Verify all books displaying contain search term

Test was triggered successfully on Jenkins but it failed since a result didnt contain the provided keyword.
![image](https://github.com/user-attachments/assets/6518da07-f99e-4220-a3ca-02caab70cbeb)

catssaymeow committed:
PACTH-6: Add Shadow DOM test

nbmhoang commented on src/main/java/pages/books/SearchResultsPage.java:
```diff
@@ -0,0 +1,58 @@
+
+    @Step("Get book title from search results")
+    private List<String> getBookTitleFromSearchResults() {
+        ElementsCollection images = $$(shadowDeepCss(bookImage)).shouldHave(CollectionCondition.sizeGreaterThan(0));
```
the size > 0 condition is not necessary here, what happen if users search book and got no result?

yenleagest replied:
It’s necessary. Without it, the initial list would be empty. Also, your assumption that the user might get no results when searching is unlikely to happen — results are always shown, even when the user enters a random string.

![image](https://github.com/user-attachments/assets/13c24ce1-c758-42b9-83f5-f329469f70d3)

nbmhoang commented on src/main/java/pages/books/SearchResultsPage.java:
```diff
@@ -0,0 +1,58 @@
+        // if the number of results does not increase after scrolling, increase retry count
+        // if new results appear, reset retry and re-evaluate until max retries is reached
+        int retry = 0;
+        while (retry < 3) {
```
What are you doing here? 🤔 Are you supposed to handle lazy loading on this page?

yenleagest replied:
Yes its a handle for lazy load. Please read my comment above the method.

catssaymeow committed:
Fix review comments

yenleagest commented:
Hi @nbmhoang, I’ve addressed your comments - please have another look.
Thanks for taking the time to explain lazy loading and for your advice on improving my code to catch bugs more effectively.
